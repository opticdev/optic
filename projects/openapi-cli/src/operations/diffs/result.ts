import { OpenAPIV3 } from '../../specs';

export enum OperationDiffResultKind {
  // Operation
  UnmatchedPath = 'UnmatchedPath',
  UnmatchedPathParameter = 'UnmatchedPathParameter',
  UnmatchedMethod = 'UnmatchedMethod',

  // Request Body
  MissingRequestBody = 'MissingRequestBody',
  UnmatchedRequestBody = 'UnmatchedRequestBody',

  // Response Body
  MissingResponseBody = 'MissingResponseBody',
  UnmatchedResponseStatusCode = 'UnmatchedResponseStatusCode',
  UnmatchedResponseBody = 'UnmatchdResponseBody',
}

export type OperationDiffResult = {} & (
  | {
      kind: OperationDiffResultKind.UnmatchedPath;
      subject: string;
    }
  | {
      kind: OperationDiffResultKind.UnmatchedPathParameter;
      subject: string;
    }
  | {
      kind: OperationDiffResultKind.UnmatchedMethod;
      subject: OpenAPIV3.HttpMethods;
      pathPattern: string;
    }
  | {
      kind: OperationDiffResultKind.UnmatchedRequestBody;
      contentType: string | null;
    }
  | {
      kind: OperationDiffResultKind.MissingRequestBody;
    }
  | {
      kind: OperationDiffResultKind.UnmatchedResponseStatusCode;
      contentType: string | null;
      statusCode: string;
    }
  | {
      kind: OperationDiffResultKind.UnmatchedResponseBody;
      contentType: string | null;
      statusCode: string;
    }
  | {
      kind: OperationDiffResultKind.MissingResponseBody;
      statusCode: string;
    }
);
