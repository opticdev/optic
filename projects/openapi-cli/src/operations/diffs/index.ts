import { Operation } from '..';
import { OpenAPIV3 } from '../../specs';
import { CapturedInteraction } from '../../captures';
import { OperationDiffResult, OperationDiffResultKind } from './result';
import {
  OperationInteractionDiffTraverser,
  SpecOperationDiffTraverser,
} from './traversers';

export type { OperationDiffResult };
export { OperationDiffResultKind };

export function* diffInteractionByOperation(
  interaction: CapturedInteraction,
  operation: Operation
): IterableIterator<OperationDiffResult> {
  const traverser = new OperationInteractionDiffTraverser();
  traverser.traverse(interaction, operation);
  yield* traverser.results();
}

export function* diffOperationWithSpec(
  operation: {
    pathPattern: string;
    methods: OpenAPIV3.HttpMethods[];
  },
  spec: OpenAPIV3.Document
): IterableIterator<OperationDiffResult> {
  const traverser = new SpecOperationDiffTraverser();
  traverser.traverse(operation, spec);
  yield* traverser.results();
}
