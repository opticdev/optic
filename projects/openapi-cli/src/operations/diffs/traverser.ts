import { CapturedInteraction } from '../../captures';
import { Operation } from '..';
import { OperationDiffResult } from './result';
import { OpenAPIV3 } from '../../specs';
import invariant from 'ts-invariant';

import { visitRequestBody, visitResponses } from './visitors';

export class OperationDiffTraverser {
  private interaction?: CapturedInteraction;
  private operation?: Operation;

  traverse(interaction, operation) {
    this.interaction = interaction;
    this.operation = operation;
  }

  *results(): IterableIterator<OperationDiffResult> {
    if (!this.interaction || !this.operation) return;

    const { operation, interaction } = this;

    const requestBody = operation.requestBody;
    invariant(
      !requestBody || isNotReferenceObject(requestBody),
      `operation expected to not have any references, found in request body, pathPattern=${operation.pathPattern} method=${operation.method}`
    );
    yield* visitRequestBody(interaction.request, requestBody);

    const responses = Object.fromEntries(
      Object.entries(operation.responses).map(([code, response]) => {
        invariant(
          isNotReferenceObject(response),
          `operation expected to not have any reference, found in response, statusCode=${code} pathPattern=${operation.pathPattern} method=${operation.method}`
        );
        return [code, response];
      })
    );
    yield* visitResponses(interaction.response, responses);
  }
}

const isNotReferenceObject = <T extends {}>(
  maybeReference: T | OpenAPIV3.ReferenceObject
): maybeReference is Exclude<T, OpenAPIV3.ReferenceObject> => {
  return !('$ref' in maybeReference);
};
