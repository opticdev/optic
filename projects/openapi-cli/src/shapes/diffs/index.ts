import { ShapeDiffTraverser } from './traverser';
import { ShapeDiffResult, ShapeDiffResultKind } from './result';
import { Body, SchemaObject } from '../body';

export type { ShapeDiffResult, ShapeDiffResultKind };

export function* diffBodyBySchema(
  body: Body,
  schema: SchemaObject
): IterableIterator<ShapeDiffResult> {
  let traverser = new ShapeDiffTraverser();
  traverser.traverse(body, schema);
  yield* traverser.results();
}
