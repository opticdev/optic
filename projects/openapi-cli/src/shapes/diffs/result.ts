import { FieldLocation } from '@useoptic/openapi-utilities';
import { JsonPath } from '@useoptic/openapi-io';
import { JsonSchemaKnownKeyword } from './traverser';

export enum ShapeDiffResultKind {
  AdditionalProperty = 'AdditionalProperty',
  MissingRequiredProperty = 'MissingRequiredProperty',
  UnmatchedType = 'UnmatchedType',
}

// The result of matching a body against it's (JSON) schema
export type ShapeDiffResult = {
  keyword: JsonSchemaKnownKeyword;

  // TODO: figure out if location belongs here, or whether one level up (its just passed
  // in and attached, not needed for logic)
  // location: FieldLocation;
  key: string;
  example: any;
  instancePath: JsonPath;
} & (
  | {
      kind: ShapeDiffResultKind.AdditionalProperty;
      keyword: JsonSchemaKnownKeyword.additionalProperties;

      propertyExamplePath: JsonPath;
      parentObjectPath: JsonPath;
      propertyPath: JsonPath;
    }
  | {
      kind: ShapeDiffResultKind.MissingRequiredProperty;
      keyword: JsonSchemaKnownKeyword.required;

      parentObjectPath: JsonPath;
      propertyPath: JsonPath;
    }
  | {
      kind: ShapeDiffResultKind.UnmatchedType;
      keyword: JsonSchemaKnownKeyword.type | JsonSchemaKnownKeyword.oneOf;

      propertyPath: JsonPath;
    }
);
