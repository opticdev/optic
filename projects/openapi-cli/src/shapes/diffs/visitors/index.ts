import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { JsonSchemaKnownKeyword } from '..';
import { ShapeDiffResult } from '../result';
import { ErrorObject } from '../traverser';

import { additionalProperties } from './additionalProperties';
import { oneOfKeyword } from './oneOf';
import { requiredKeyword } from './required';
import { typeKeyword } from './type';

export interface ShapeDiffVisitor {
  (
    validationError: ErrorObject,
    example: any
  ): IterableIterator<ShapeDiffResult>;
}

const visitors: ShapeDiffVisitor[] = [
  additionalProperties,
  oneOfKeyword,
  requiredKeyword,
  typeKeyword,
];

export function* diffVisitors(
  validationError: ErrorObject,
  example: any
): IterableIterator<ShapeDiffResult> {
  for (let visitor of visitors) {
    yield* visitor(validationError, example);
  }
}
