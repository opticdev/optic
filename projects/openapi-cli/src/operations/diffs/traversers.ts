import { CapturedInteraction } from '../../captures';
import { Operation } from '..';
import { OperationDiffResult } from './result';
import { OperationQueries } from '../queries';
import { OpenAPIV3, SpecFactsIterable } from '../../specs';
import { visitRequestBody, visitResponses } from './visitors/index';
import { visitPath, visitMethod } from './visitors/index';
import { Some, None } from 'ts-results';

export class OperationInteractionDiffTraverser {
  private interaction?: CapturedInteraction;
  private operation?: Operation;

  traverse(interaction, operation) {
    this.interaction = interaction;
    this.operation = operation;
  }

  *results(): IterableIterator<OperationDiffResult> {
    if (!this.interaction || !this.operation) return;

    const { operation, interaction } = this;

    yield* visitRequestBody(interaction.request, operation.requestBody);

    yield* visitResponses(interaction.response, operation.responses);
  }
}

export class SpecOperationDiffTraverser {
  private operation?: {
    pathPattern: string;
    methods: OpenAPIV3.HttpMethods[];
  };
  private spec?: OpenAPIV3.Document;
  private queries?: OperationQueries;

  traverse(operation, spec) {
    this.operation = operation;
    this.spec = spec;
    // TODO: figure out whether the cost of rebuilding queries from facts for each
    // traversal is acceptable
    let facts = SpecFactsIterable.fromOpenAPISpec(spec);
    this.queries = OperationQueries.fromFacts(facts);
  }

  *results(): IterableIterator<OperationDiffResult> {
    if (!this.operation || !this.queries || !this.spec) return;
    const { operation, queries, spec } = this;

    let pathPatternOption = queries
      .findPathPattern(operation.pathPattern)
      .expect('path pattern should be able to be matched against spec');

    let pathSpecOption = pathPatternOption.map(
      (pathPattern): OpenAPIV3.PathItemObject => spec.paths[pathPattern]!
    );

    yield* visitPath(operation.pathPattern, pathSpecOption, {
      pathPattern: pathPatternOption,
    });

    if (pathSpecOption.some && pathPatternOption.some) {
      let pathSpec = pathSpecOption.val;
      let pathPattern = pathPatternOption.val;

      for (let method of operation.methods) {
        let methodSpec = pathSpec[method];

        yield* visitMethod(method, methodSpec ? Some(methodSpec) : None, {
          pathPattern,
        });
      }
    }
  }
}
