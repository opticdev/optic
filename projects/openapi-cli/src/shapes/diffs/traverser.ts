import jsonSchemaTraverse from 'json-schema-traverse';
import { SchemaObject, Body } from '..';
import { ShapeDiffResult } from './result';
import Ajv, { ErrorObject, ValidateFunction } from 'ajv';
import { OpenAPIV3 } from '../../specs';
import { jsonPointerHelpers } from '@useoptic/json-pointer-helpers';
import { Ono } from '@jsdevtools/ono';
import { Result, Ok, Err } from 'ts-results';

import { diffVisitors } from './visitors';

export type { ErrorObject };

export class ShapeDiffTraverser {
  private validator: Ajv;

  private validate?: ValidateFunction;
  private bodyValue?: any;

  constructor() {
    this.validator = new Ajv({
      allErrors: true,
      validateFormats: false,
      strictSchema: false,
      strictTypes: false,
      useDefaults: true,
    });
  }

  traverse(
    bodyValue: any,
    schema: SchemaObject
  ): Result<void, SchemaCompilationError> {
    this.bodyValue = bodyValue;
    try {
      this.validate = this.validator.compile(prepareSchemaForDiff(schema));
    } catch (err) {
      // Catching and not throwing is okay here. `validator.compile` is stateless, with all state for the validator
      // being contained by the function that now doesn't get assigned.
      const wrapped = new SchemaCompilationError(err as Error);
      return Err(wrapped);
    }

    this.validate(bodyValue);

    return Ok.EMPTY;
  }

  *results(): IterableIterator<ShapeDiffResult> {
    if (!this.validate || !this.validate.errors) return;
    let validationErrors = this.validate.errors;

    let oneOfs: Map<string, ErrorObject> = new Map();
    let oneOfBranchType: [string, ErrorObject][] = [];
    let oneOfBranchOther: [string, ErrorObject][] = [];

    for (let validationError of validationErrors) {
      if (validationError.keyword === JsonSchemaKnownKeyword.oneOf) {
        let schemaPath = validationError.schemaPath.substring(1); // valid json pointer
        oneOfs.set(schemaPath, validationError);
      } else if (validationError.schemaPath.indexOf('oneOf') > -1) {
        // probably has a oneof ancestor

        let schemaPath = jsonPointerHelpers.decode(
          validationError.schemaPath.substring(1)
        );
        let oneOfPath = schemaPath.slice(
          0,
          schemaPath.lastIndexOf('oneOf') + 1
        );
        let branchPath = schemaPath.slice(oneOfPath.length);

        // TODO: consider hardening detection of one of branches. Going just off string keynames could
        // potentially be ugly, as we're dealing with partially user-definable input (property names).
        if (
          (branchPath.length == 2 &&
            branchPath[1] === JsonSchemaKnownKeyword.type) ||
          (branchPath.length === 3 &&
            branchPath[1] === 'items' &&
            branchPath[2] === JsonSchemaKnownKeyword.type)
        ) {
          oneOfBranchType.push([
            jsonPointerHelpers.compile(oneOfPath),
            validationError,
          ]);
        } else if (branchPath.length >= 2) {
          oneOfBranchOther.push([
            jsonPointerHelpers.compile(oneOfPath),
            validationError,
          ]);
        }
      } else {
        // not related to one-of? visit right away
        yield* diffVisitors(validationError, this.bodyValue);
      }
    }

    for (let [oneOfPath, otherBranchError] of oneOfBranchOther) {
      // any nested errors are all safe to visit
      yield* diffVisitors(otherBranchError, this.bodyValue);

      // once a nested error has been visited, we consider this a branch type match
      oneOfs.delete(oneOfPath);
      oneOfBranchType = oneOfBranchType.filter(
        ([branchOneOfPath, _]) => oneOfPath !== branchOneOfPath
      );
    }

    // visit any left over one ofs
    for (let oneOfError of oneOfs.values()) {
      yield* diffVisitors(oneOfError, this.bodyValue);
    }
  }
}

export enum JsonSchemaKnownKeyword {
  required = 'required',
  additionalProperties = 'additionalProperties',
  type = 'type',
  oneOf = 'oneOf',
}

export class SchemaCompilationError extends Error {
  constructor(ajvError: Error) {
    super(`Error compiling schema: ${ajvError.message}`);
    Ono.extend(this, ajvError);
    Object.setPrototypeOf(this, SchemaCompilationError.prototype);
  }
}

/*
  Some developers don't set all the JSON Schema properties because it's quite verbose,
  effectively underspecifing their schemas. Optic tries to apply sensible defaults,
  which are easy to override by writing your schemas properly
 */
function prepareSchemaForDiff(input: SchemaObject): SchemaObject {
  const schema: SchemaObject = JSON.parse(JSON.stringify(input));
  jsonSchemaTraverse(schema as OpenAPIV3.SchemaObject, {
    allKeys: true,
    cb: (schema) => {
      if (
        schema.type === 'object' &&
        !schema.hasOwnProperty('additionalProperties')
      ) {
        schema['additionalProperties'] = false;
      }
    },
  });
  return schema;
}
