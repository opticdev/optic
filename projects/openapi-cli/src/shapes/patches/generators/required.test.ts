import { it, describe, expect } from '@jest/globals';
import { SchemaObject } from '../..';
import { diffValueBySchema } from '../../diffs';
import { generateShapePatchesByDiff } from '..';

describe('required shape patch generator', () => {
  const jsonSchema: SchemaObject = {
    type: 'object',
    properties: {
      hello: {
        type: 'object',
        required: ['f1', 'f2', 'f3'],
        properties: {
          f1: { type: 'string' },
          f2: { type: 'number' },
          f3: { type: 'object' },
        },
      },
    },
  };

  it("missing required fields will patch as 'make optional' or 'remove'", () => {
    const input = {
      hello: { f1: 'value', f2: 122 },
    };
    const diffs = [...diffValueBySchema(input, jsonSchema)];

    const patches = diffs.flatMap((diff) => [
      ...generateShapePatchesByDiff(diff, jsonSchema, {}),
    ]);

    expect(patches).toMatchSnapshot();
  });
});
