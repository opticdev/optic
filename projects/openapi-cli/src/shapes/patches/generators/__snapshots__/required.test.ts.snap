// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`required shape patch generator missing required fields will patch as 'make optional' or 'remove' 1`] = `
[
  {
    "description": "make property f3 optional",
    "diff": {
      "description": "required property 'f3' was missing",
      "example": undefined,
      "instancePath": "/hello/f3",
      "key": "f3",
      "keyword": "required",
      "kind": "MissingRequiredProperty",
      "parentObjectPath": "/properties/hello",
      "propertyPath": "/properties/hello/properties/f3",
    },
    "groupedOperations": [
      {
        "intent": "remove f3 from parent's required array",
        "operations": [
          {
            "op": "remove",
            "path": "/properties/hello/required/2",
          },
        ],
      },
    ],
    "impact": [
      "Addition",
      "BackwardsCompatibilityUnknown",
    ],
  },
  {
    "description": "remove property f3",
    "diff": {
      "description": "required property 'f3' was missing",
      "example": undefined,
      "instancePath": "/hello/f3",
      "key": "f3",
      "keyword": "required",
      "kind": "MissingRequiredProperty",
      "parentObjectPath": "/properties/hello",
      "propertyPath": "/properties/hello/properties/f3",
    },
    "groupedOperations": [
      {
        "intent": "remove f3 from parent's required array",
        "operations": [
          {
            "op": "remove",
            "path": "/properties/hello/required/2",
          },
        ],
      },
      {
        "intent": "remove f3 from parent's properties object",
        "operations": [
          {
            "op": "remove",
            "path": "/properties/hello/properties/f3",
          },
        ],
      },
    ],
    "impact": [
      "BackwardsCompatibilityUnknown",
    ],
  },
]
`;
